File: backend/domain/src/main/java/com/ljwzz/weathertrafficalarm/backend/domain/model/PlaceRef.java
package com.ljwzz.weathertrafficalarm.backend.domain.model;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

public record PlaceRef(
    String poiId,
    @NotBlank String name,
    @NotBlank String displayAddress,
    double longitudeGcj02,
    double latitudeGcj02,
    @NotBlank String adcode,
    @NotBlank String citycode
) {
    public PlaceRef {
        if (longitudeGcj02 < -180 || longitudeGcj02 > 180) {
            throw new IllegalArgumentException("longitude out of range: " + longitudeGcj02);
        }
        if (latitudeGcj02 < -180 || latitudeGcj02 > 180) {
            throw new IllegalArgumentException("latitude out of range: " + latitudeGcj02);
        }
    }
}

File: android/core/model/src/main/java/com/ljwzz/weathertrafficalarm/core/model/PlaceRef.kt
package com.ljwzz.weathertrafficalarm.core.model

import kotlinx.serialization.Serializable

@Serializable
data class PlaceRef(
    val poiId: String? = null,
    val name: String,
    val displayAddress: String,
    val longitudeGcj02: Double,
    val latitudeGcj02: Double,
    val adcode: String,
    val citycode: String,
) {
    init {
        require(longitudeGcj02 in -180.0..180.0) { "longitude out of range: $longitudeGcj02" }
        require(latitudeGcj02 in -180.0..180.0) { "latitude out of range: $latitudeGcj02" }
        require(name.isNotBlank()) { "name must not be blank" }
    }

    override fun toString(): String =
        "PlaceRef(name=$name, displayAddress=$displayAddress, poiId=$poiId, adcode=$adcode, citycode=$citycode)"
}
